### PR TITLE
construct symbol for converting ZF net

### PR DESCRIPTION
**Pull request by liangfu**
constructed symbol and converted ZFnet caffemodel for fine-tuning with faster rcnn;

### CODE REVIEW

**Comment by piiswrong** [comment]
@precedenceguo

**Comment by liangfu** [comment]
this PR is not yet ready to merge into `master` branch

**Comment by precedenceguo** [comment]
The converter is the same as vgg so I guess the model converted should work. Well no one ever reported the zf imagenet model accuracy. So without imagenet pretrained model, making training work could be tough.

**Comment by liangfu** [comment]
pretrained zf model are available at https://github.com/ShaoqingRen/faster_rcnn, and **RPN+ZF, shared** model with 300 proposals at test time result in 59.9% mAP on PASCAL VOC 2007 test set, as stated on the original FasterRCNN article.

**Comment by precedenceguo** [comment]
a) What about the imagenet top-1 precision of pretrained model?
b) Does the updated code reproduce this 59.9 mAP?

**Comment by liangfu** [comment]
a) As the the pretrained model is optimized for object detection task, it is not evaluated on imagenet classification task
b) I can't reproduce 59.9 mAP on pascal voc 2007 data set at current stage, as the LRN layer in mxnet and caffe behaves quite different (I've opened #4317 to track this issue). So there is still more room to improve, and I would continue to push to this branch when I'm getting close to the reported mAP.

**Comment by piiswrong** [comment]
any updates on this?

**Comment by piiswrong** [comment]
Can one of the admins verify this patch?

**Comment by mli** [comment]
closed since there is no progress. please feel free to reopen it

**Comment by DengpanFu** [comment]
I have done the same thing with ZF-net too, and finally i got 60.89 mAP on voc_2007_test  with model trained with voc_2007_trainvel and 67.01 with voc_2007_trainval+voc_2012_trainval. I just ignored the difference of LRN between mxnet and caffe, but the important thing is that, you should finetune all the layers in ZF(while VGG/Resnet was trained with conv1 and conv2 fixed).

**Comment by liangfu** [comment]
@DengpanFu Thanks for sharing the interesting result. Are you interested in creating a PR so that everyone can reproduce that?

**Comment by tianzening** [comment]
@liangfu Thanks for your work.  when I use your convert_zf.py,https://github.com/apache/incubator-mxnet/pull/4221/commits/fcd1249e8d5aec0ac1c7dda704cd07c0b0b3eab7
 I face some problems. Can you guide me how to modify the test.prototxt. （I want to use the py-faster-rcnn/models/pascall_voc/ZF/faster_rcnn_end2end/test.prototxt, and pretraind ZF.v2.caffemodel to generate ZF-0000.params ）.

Error:  Message type "caffe.LayerParameter" has no field named "roi_pooling_param"

#========= RCNN ============
# layer {
#   name: "roi_pool_conv5"
#   type: "ROIPooling"
#   bottom: "conv5"
#   bottom: "rois"
#   top: "roi_pool_conv5"
#   # roi_pooling_param {
#   #   pooled_w: 6
#   #   pooled_h: 6
#   #   spatial_scale: 0.0625 # 1/16
#   # }
# }
Then I comment the above part，Now  there is another error。

Cannot copy param 0 weights from layer 'fc6'; shape mismatch.  Source param shape is 4096 9216 (37748736); target param shape is 4096 57600 (235929600). To learn this layer's parameters from scratch rather than copying from a saved net, rename the layer.

Thanks

**Comment by liangfu** [comment]
@tianzening  As stated at https://github.com/rbgirshick/fast-rcnn/issues/130 :
"It looks like you're trying to use a version of Caffe not included in fast-rcnn because your missing layers that aren't included in the main branch; you should try reinstalling using the version of caffe in the fast-rcnn branch."
